File: src/main/java/com/mynimef/bot/config/BotCreator.java
package com.mynimef.bot.config;

import com.mynimef.bot.IBot;
import com.mynimef.bot.actions.ISaveLogs;
import org.telegram.telegrambots.meta.TelegramBotsApi;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException;
import org.telegram.telegrambots.updatesreceivers.DefaultBotSession;


public final class BotCreator {

    private final String token;
    private final String username;

    private CommandsBuilder commands;
    private CallbacksBuilder callbacks;

    private ISaveLogs logs;

    public BotCreator(
            String token,
            String username
    ) {
        this.token = token;
        this.username = username;
    }

    public BotCreator addCommands(CommandsBuilder commands) {
        this.commands = commands;
        return this;
    }

    public BotCreator addCallbacks(CallbacksBuilder callbacks) {
        this.callbacks = callbacks;
        return this;
    }

    public BotCreator addLogs(ISaveLogs logs) {
        this.logs = logs;
        return this;
    }

    public IBot start() {
        if (commands == null) {
            commands = getEmptyCommands();
        }

        return initBot(
                new TelegramBot(
                        commands.getCommands(),
                        commands.getNoCommandRecognized(),
                        (callbacks != null) ? callbacks.getCallbacks() : null,
                        token,
                        username,
                        logs
                )
        );
    }

    private TelegramBot initBot(TelegramBot bot) {
        try {
            TelegramBotsApi telegramBotsApi = new TelegramBotsApi(DefaultBotSession.class);

            try {
                telegramBotsApi.registerBot(bot);
                return bot;
            } catch (TelegramApiRequestException e) {
                e.printStackTrace();
            }
        } catch (TelegramApiException e) {
            e.printStackTrace();
        }

        return null;
    }

    private CommandsBuilder getEmptyCommands() {
        return new CommandsBuilder() {
            @Override
            public void init() { }

            @Override
            protected void nonCommandUpdate(
                    String text,
                    String chatId,
                    Integer messageId,
                    String username,
                    String firstName,
                    String lastName,
                    IBot bot
            ) {
                bot.sendMessage(chatId, "Configure CommandsBuilder class");
            }
        };
    }
}

File: src/main/java/com/mynimef/bot/config/BotCreator.kt
package com.mynimef.bot.config

import com.mynimef.bot.IBot
import org.telegram.telegrambots.meta.TelegramBotsApi
import org.telegram.telegrambots.meta.exceptions.TelegramApiException
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException
import org.telegram.telegrambots.updatesreceivers.DefaultBotSession

class BotCreator(
    private val token: String,
    private val username: String
) {

    private var commands: CommandsBuilder? = null
    private var callbacks: CallbacksBuilder? = null

    private var logs: SaveLog? = null

    fun addCommands(commands: CommandsBuilder?): BotCreator {
        this.commands = commands
        return this
    }

    fun addCallbacks(callbacks: CallbacksBuilder?): BotCreator {
        this.callbacks = callbacks
        return this
    }

    fun addLogs(logs: SaveLog): BotCreator {
        this.logs = logs
        return this
    }

    fun start(): IBot? {
        if (commands == null) {
            commands = object : CommandsBuilder() {
                public override fun init() {}
                override val nonCommandUpdate: Action
                    get() = { it, bot ->
                        bot.sendMessage(it.chatId, "Configure CommandsBuilder class")
                    }
            }
        }

        return initBot(
            TelegramBot(
                commands = commands!!.commands,
                noCommandRecognized = commands!!.nonCommandUpdate,
                callbacks = if ((callbacks != null)) callbacks!!.callbacks else null,
                token = token,
                username = username,
                saveLog = logs
            )
        )
    }

    private fun initBot(bot: TelegramBot): TelegramBot? {
        try {
            val telegramBotsApi = TelegramBotsApi(DefaultBotSession::class.java)

            try {
                telegramBotsApi.registerBot(bot)
                return bot
            } catch (e: TelegramApiRequestException) {
                e.printStackTrace()
            }
        } catch (e: TelegramApiException) {
            e.printStackTrace()
        }

        return null
    }

}
